src/Diagnostics.DataProviders/DataProviders.cs

```diff
@@ -18,27 +20,85 @@ public class DataProviders
         public Func<MdmDataSource, IMdmDataProvider> Mdm;
         public Func<GenericMdmDataProviderConfiguration, IMdmDataProvider> MdmGeneric;
 
+        private readonly Dictionary<object, IMetadataProvider> _dataProvidersCache = new Dictionary<object, IMetadataProvider>();
+
         public DataProviders(DataProviderContext context)
         {
-            Kusto = new KustoLogDecorator(context, new KustoDataProvider(_cache, context.Configuration.KustoConfiguration, context.RequestId, context.KustoHeartBeatService));
-            Observer = new ObserverLogDecorator(context, SupportObserverDataProviderFactory.GetDataProvider(_cache, context.Configuration, context));
-            GeoMaster = new GeoMasterLogDecorator(context, new GeoMasterDataProvider(_cache, context));
-            AppInsights = new AppInsightsLogDecorator(context, new AppInsightsDataProvider(_cache, context.Configuration.AppInsightsConfiguration));
-            ChangeAnalysis = new ChangeAnalysisLogDecorator(context, new ChangeAnalysisDataProvider(_cache, context.Configuration.ChangeAnalysisDataProviderConfiguration, context.RequestId, context.clientObjectId, context.clientPrincipalName, Kusto, context.receivedHeaders));
-            Asc = new AscLogDecorator(context, new AscDataProvider(_cache, context.Configuration.AscDataProviderConfiguration, context.RequestId, context));
-            Mdm = (MdmDataSource ds) =>
+
+            Kusto = FromMemoryCache("Kusto", context) as IKustoDataProvider;
+            Observer = FromMemoryCache("Observer", context) as ISupportObserverDataProvider;
+            GeoMaster = FromMemoryCache("GeoMaster", context) as IGeoMasterDataProvider;
+            AppInsights = FromMemoryCache("AppInsights", context) as IAppInsightsDataProvider;
+            ChangeAnalysis = FromMemoryCache("ChangeAnalysis", context) as IChangeAnalysisDataProvider;
+            Asc = FromMemoryCache("Asc", context) as IAscDataProvider;
+            Mdm = (MdmDataSource ds) => FromMemoryCache(ds, context) as IMdmDataProvider;
+            MdmGeneric = (GenericMdmDataProviderConfiguration config) => FromMemoryCache(config, context) as IMdmDataProvider;
+        }
+
+        private IMetadataProvider FromMemoryCache(object key, DataProviderContext context)
```

Comment:
Hey Puneet, maybe use this instead. This way you don't have to have a huge switch case statement and maintain a static list of data provider names.

        private T GetOrAddDataProvider<T>(T dataProvider) where T : LogDecoratorBase
        {
            _dataProviderList.Add(dataProvider);
            return dataProvider;
        }